### PR TITLE
Dev/add pytorch descriptor generator

### PR DESCRIPTION
**Pull request by joshanderson-kw**
Adds pytorch descriptor generator implementation based off of DIY AI's implementation

### CODE REVIEW

**Comment by joshanderson-kw** [comment]
I think the remaining ghostflow errors are from commits made in the other repo.

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+    Array/Matrix normalization along max dimension (i.e. a=0 for 1D array, a=1
+    for 2D array, etc.).
+
+    :param np.ndarray v:
+        Vector to normalize.
+    :param None|int|float|str mode:
+        ``numpy.linalg.norm`` order parameter.
+
+    :return: Normalize version of the input array ``v``.
+    :rtype: np.ndarray
```
Comment (Purg) [inline smqtk_descriptors/impls/descriptor_generator/pytorch.py:58]: The type and `:rtype:` parts may be removed: type annotations should supersede them.

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+__all__ = [
+    "TorchModuleDescriptorGenerator",
+    "Resnet50SequentialTorchDescriptorGenerator",
+    "AlignedReIDResNet50TorchDescriptorGenerator"
+]
+T = TypeVar("T", bound="TorchModuleDescriptorGenerator")
+
+
+def normalize_vectors(v: np.ndarray,
+                      mode: Optional[Union[int, float, str]] = None) -> np.ndarray:
```
Comment (Purg) [inline smqtk_descriptors/impls/descriptor_generator/pytorch.py:47]: We can alternatively split up long parameter lines like:
```suggestion
def normalize_vectors(
    v: np.ndarray,
    mode: Optional[Union[int, float, str]] = None
) -> np.ndarray:
```

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+    # When normalization off
+    return v
+
+
+try:
+    class ImgMatDataset (Dataset):
+
+        def __init__(self,
+                     img_mat_list: Sequence[np.ndarray],
+                     transform: Callable[[Iterable[np.ndarray]], "torch.Tensor"]) -> None:
```
Comment (Purg) [inline smqtk_descriptors/impls/descriptor_generator/pytorch.py:74]: We can alternatively  split up long parameter lines like:
```suggestion
        def __init__(
            self,
            img_mat_list: Sequence[np.ndarray],
            transform: Callable[[Iterable[np.ndarray]], "torch.Tensor"]
        ) -> None:
```

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+        latency to first yield.
+        This mode is idea for streaming input or high input volume situations.
+        This mode currently has a short-coming of working only in a threaded
+        manner so it is not as fast as using the `DataLoader` avenue.
+    :param bool global_average_pool:
+        Optionally apply a GAP operation to the spatial dimension of the feature
+        vector that is returned by the descriptor generator. Some models return
+        a w x h x k tensor and flatten them into a single dimension, which can
+        remove the spatial context of the features. Taking an average over each
+        channel can improve the clustering in some cases.
```
Comment (Purg) [inline smqtk_descriptors/impls/descriptor_generator/pytorch.py:133]: We can remove the parameter types here as the type annotations should supersede these.

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+    def from_config(
+        cls: Type[T],
+        config_dict: Dict,
+        merge_default: bool = True
+    ) -> T:
+        # Copy config to prevent input modification
+        config_dict = copy.deepcopy(config_dict)
+        config_dict['image_reader'] = \
+            from_config_dict(config_dict['image_reader'],
+                             ImageReader.get_impls())
```
Comment (Purg) [inline smqtk_descriptors/impls/descriptor_generator/pytorch.py:159]: In an effort to reduce the use of `\`s, we can reformat this to
```suggestion
        config_dict['image_reader'] = from_config_dict(
            config_dict['image_reader'],
            ImageReader.get_impls()
        )
```

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+        iter_runtime: bool = False,
+        global_average_pool: bool = False
+    ):
+        super().__init__()
+
+        # Make sure given image reader is valid
+        try:
+            image_reader.is_valid_element
+        except AttributeError:
+            raise
```
Comment (Purg) [inline smqtk_descriptors/impls/descriptor_generator/pytorch.py:181]: I'm not sure I understand what this block is for. I don't see it in the source we are drawing from, and I don't know what this is doing: the except catches an exception, does nothing with it and then just re-raises it. I either suggest we don't need this block or this needs more explanation comments on why this is here.

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py (hunk 1 above)
Comment (Purg) [reply smqtk_descriptors/impls/descriptor_generator/pytorch.py:58]: Actually, looks like this is a theme here. Other functions below could do with some de-duplication between type annotations and doc-string parameter docs.

**Review by Purg — COMMENTED** [review]
I get a "The lock file is not up to date" warning from poetry when attempting to do a local development install of the updated dependencies. This usually indicates that a `poetry lock --no-update` is required to sync the files.

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py [hunk 7]
```diff
@@ -175,10 +174,7 @@ def __init__(
         super().__init__()
 
         # Make sure given image reader is valid
-        try:
-            image_reader.is_valid_element
-        except AttributeError:
-            raise
+        assert image_reader is not None
```
Comment (Purg) [inline smqtk_descriptors/impls/descriptor_generator/pytorch.py:177]: I guess to rephrase my previous question: why are we checking the input `image_reader` value at all? I don't understand or am missing what the failure-case is that we are protecting against here.

File: smqtk_descriptors/impls/descriptor_generator/pytorch.py (hunk 6 above)
Comment (joshanderson-kw) [reply smqtk_descriptors/impls/descriptor_generator/pytorch.py]: I had added it to make sure that the provided image_reader parameter is valid. Tested in `test_init_no_image_reader`, which is based on the `test_init_no_*` tests in test_caffe.py. I went ahead and changed it to `assert image_reader is not None`, so hopefully it's a bit more clear now.

**Review by Purg — APPROVED** [review]

**Pull request merged**